### ISSUE TITLE
Exploding gradients through convTranspose layer

### ISSUE THREAD

**Issue by arunmallya** [vlfeat/matconvnet#283]
I was trying to finetune the pascal-fcn32s-dag network by attaching a softmaxloss (dagnn.Loss) layer at the end and then backpropagating wrt to the correct label. The loss explodes and params reach NaN after processing 2-3 images. 

After a bit of investigation, it looks like the gradients passed down from the convT (upsample) layer are huge, with an elementwise max/min over the derivative having a magnitude of 500+ for the first input image, with further deterioration after param updates.
The forward pass works totally fine, but the backprop breaks.

Furthermore, if I clear the params of the convT layer and re-initialize them, the same backprop issue pops up immediately.

Any ideas as to why that might be happening?

**Comment by vedaldi** [comment]
Hi, you probably need a little tuning. I experienced similar issues myself when implementing training of FCN and could solve them by changing the initialisation and learning rate for a few layers. At any rate, you may want to have a look at our just-released implementation here:

https://github.com/vlfeat/matconvnet-fcn <https://github.com/vlfeat/matconvnet-fcn>

**Comment by arunmallya** [comment]
Thanks! Will check it out.

**Comment by arunmallya** [comment]
Is there any reason the gradients explode if the deconv is not set to bilinear interpolation? If I initialize deconv to random weights, with a non-zero learning rate, it still trains well in caffe, while things go south very quickly in matconvnet.

**Comment by vedaldi** [comment]
Hi, there shouldn’t be any particular reason, no.

There could be small but important practical differences, however. For example, the learning rate, or the norm of the initial random weights. Perhaps Caffe is normalising them implicitly during initialisation?

For some odd reasons, networks like VGG VD tend to have pretty large dynamic range for their features; so if you attach on top a classification layer with sufficiently large random weights, it may be break. My suggestion would be to try to initialise with smaller random weights (e.g. divide them by 10 or a 100), or make their “norm” similar to the bilinear weights, and/or reduce the learning rate by a factor of 10 for those particular layers.

Of course, a bug is always possible, but I would be surprised as the reset seems to work fine.

**Comment by clabaus** [comment]
@arunmallya  I am facing similar problems, even with really small learning rates. Have you been able to solve this problem?
